packages/payment-detection/src/reference-based-detector.ts

```diff
@@ -0,0 +1,224 @@
+  ): Promise<ExtensionTypes.IAction> {
+    // If no salt is given, generate one
+    const salt =
+      paymentNetworkCreationParameters.salt || (await Utils.crypto.generate8randomBytes());
+
+    return this.extension.createCreationAction({
+      paymentAddress: paymentNetworkCreationParameters.paymentAddress,
+      refundAddress: paymentNetworkCreationParameters.refundAddress,
+      salt,
+    });
+  }
+
+  /**
+   * Creates the extensions data to add payment address
+   *
+   * @param parameters to add payment information
+   * @returns The extensionData object
+   */
+  public createExtensionsDataForAddPaymentInformation(
+    parameters: ExtensionTypes.PnReferenceBased.IAddPaymentAddressParameters,
+  ): ExtensionTypes.IAction {
+    return this.extension.createAddPaymentAddressAction({
+      paymentAddress: parameters.paymentAddress,
+    });
+  }
+
+  /**
+   * Creates the extensions data to add refund address
+   *
+   * @param Parameters to add refund information
+   * @returns The extensionData object
+   */
+  public createExtensionsDataForAddRefundInformation(
+    parameters: ExtensionTypes.PnReferenceBased.IAddRefundAddressParameters,
+  ): ExtensionTypes.IAction {
+    return this.extension.createAddRefundAddressAction({
+      refundAddress: parameters.refundAddress,
+    });
+  }
+
+  /**
+   * Gets the balance and the payment/refund events
+   *
+   * @param request the request to check
+   * @returns the balance and the payment/refund events
+   */
+  public async getBalance(
+    request: RequestLogicTypes.IRequest,
+  ): Promise<PaymentTypes.IBalanceWithEvents<TPaymentEventParameters>> {
+    if (!request.currency.network) {
+      request.currency.network = 'mainnet';
+    }
+    const supportedNetworks = this.extension.supportedNetworks;
+    if (!supportedNetworks.includes(request.currency.network)) {
+      return getBalanceErrorObject(
+        `Payment network ${request.currency.network} not supported by ${
+          this.extensionType
+        } payment detection. Supported networks: ${supportedNetworks.join(', ')}`,
+        PaymentTypes.BALANCE_ERROR_CODE.NETWORK_NOT_SUPPORTED,
+      );
+    }
+    const paymentNetwork = request.extensions[this.extensionType];
+
+    if (!paymentNetwork) {
+      return getBalanceErrorObject(
+        `The request does not have the extension: ${this.extensionType}`,
+        PaymentTypes.BALANCE_ERROR_CODE.WRONG_EXTENSION,
+      );
+    }
+
+    try {
+      const payments = await this.extractBalanceAndEvents(
+        paymentNetwork.values.paymentAddress,
+        PaymentTypes.EVENTS_NAMES.PAYMENT,
+        request.currency.network,
+        request.requestId,
+        paymentNetwork,
+      );
+      const refunds = await this.extractBalanceAndEvents(
+        paymentNetwork.values.refundAddress,
+        PaymentTypes.EVENTS_NAMES.REFUND,
+        request.currency.network,
+        request.requestId,
+        paymentNetwork,
+      );
```

Comment:
This has been factorized a bit but the logic is untouched (based on ETH payment detection)